File: chapter10/lists_and_recursion_7.ex
defmodule Primes do
  defp span(from, to) when from > to, do: []
  defp span(from, to), do: [from | span(from + 1, to)]

  def primes(n) do
    for x <- span(2, n), Enum.all?(span(2, :math.sqrt(x)), &(rem(x, &1) != 0)), do: x
  end
end

File: chapter10/lists_and_recursion_7.clj
(ns lists-and-recursion-7)

(defn- span [from to]
  (if (> from to)
    ()
    (cons from (span (inc from) to))))

(defn primes [n]
  (for [x (span 2 n)
        :when (every? #(not (zero? (rem x %)))
                      (span 2 (Math/sqrt x)))]
    x))
